### ISSUE TITLE
Comparison authenticated encryption performance Botan <-> OpenSSL

### ISSUE THREAD

**Issue by SideChannel** [randombit/botan#969]
I've made some performance tests with Botan and OpenSSL and cannot quite explain the results:

Device under Test
- OpenSuse Leap 42.2 Linux, Skylake@4GHz, 64 GByte RAM
- Compiler gcc 6.3.0 (only gcc-6 knows about the Skylake Microarchitecture)
- Compiling OpenSSL 1.1.0e
/usr/bin/perl ./Configure linux-x86_64 -Wa,--noexecstack -march=skylake -mtune=skylake --prefix=/usr/local/ssl
make;make test;sudo make install
cd /usr/local/ssl/bin
export LD_LIBRARY_PATH=/usr/local/ssl/lib:$LD_LIBRARY_PATH
ldconfig -v
- ./openssl speed -evp aes-128-gcm   gives
aes-128-gcm     653622.78k  1564458.71k  3062810.62k  4814634.33k  6001598.46k  6109298.69k
analogously
aes-192-gcm     603774.27k  1523435.88k  2806049.19k  4191125.15k  5049832.79k  5132091.39k
aes-256-gcm     563479.16k  1448331.65k  2503655.77k  3719963.99k  4349897.39k  4412506.11k
- This gives the following benchmarking slide for OpenSSL AES-GCM with AES-NI and PCLMUL
![benchmark-openssl-aes-gcm](https://cloud.githubusercontent.com/assets/26873237/24614491/a6e6ad2e-188b-11e7-87e2-cd202494a0cf.png)
- The above slide is consistent with the performance results from Intel, for example, Shay Gueron, CFRG Meeting at EUROCRYPT 2016.

Now it comes to Botan.

unzip botan-master.zip
cd botan-master
./configure.py --amalgamation --single-amalagmation-file --prefix=/usr/local/botan

  INFO: ./configure.py invoked with options "--amalgamation --single-amalgamation-file --prefix=/usr/local/botan"
   INFO: Platform: OS="Linux" machine="x86_64" proc="x86_64"
   INFO: Guessing target OS is linux (use --os to set)
   INFO: Guessing to use compiler gcc (use --cc to set)
   INFO: Detected CPU model "Intel(R) Core(TM) i7-6700K CPU @ 4.00GHz" in /proc/cpuinfo
   INFO: Guessing target processor is a x86_64/x86_64 (use --cpu to set)
   INFO: Target is gcc-linux-x86_64-x86_64
   INFO: Found sphinx-build (use --without-sphinx to disable)
   INFO: Skipping, dependency failure - certstor_sqlite3 sessions_sqlite3
   INFO: Skipping, incompatible OS - cryptoapi_rng darwin_secrandom win32_stats
   INFO: Skipping, no enabled compression schemes - compression
   INFO: Skipping, requires external dependency - boost bzip2 lzma openssl sqlite3 tpm zlib
   INFO: Loading modules adler32 aead aes aes_ni aes_ssse3 aont asn1 auto_rng base base64 bcrypt bigint blake2 block blowfish camellia cascade cast cbc cbc_mac ccm cecpq1 certstor_sql cfb chacha chacha20poly1305 chacha_sse2 clmul cmac codec_filt comb4p crc24 crc32 cryptobox ctr curve25519 des dev_random dh dl_algo dl_group dlies dsa dyn_load eax ec_gfp ec_group ecc_key ecdh ecdsa ecgdsa ecies eckcdsa elgamal eme_oaep eme_pkcs1 eme_raw emsa1 emsa_pkcs1 emsa_pssr emsa_raw emsa_x931 entropy fd_unix ffi filters fpe_fe1 gcm gmac gost_28147 gost_3410 gost_3411 hash hash_id hex hkdf hmac hmac_drbg http_util idea idea_sse2 iso9796 kasumi kdf kdf1 kdf1_iso18033 kdf2 keccak keypair lion locking_allocator mac mce mceies md4 md5 mdx_hash mgf1 misty1 mode_pad modes mp newhope noekeon noekeon_simd numbertheory ocb ofb par_hash passhash9 pbes2 pbkdf pbkdf1 pbkdf2 pem pk_pad pkcs11 poly1305 prf_tls prf_x942 proc_walk pubkey rc4 rdrand rdrand_rng rdseed rfc3394 rfc6979 rmd160 rng rsa salsa20 seed serpent serpent_simd sessions_sql sha1 sha1_sse2 sha2_32 sha2_64 sha3 shake shake_cipher simd siphash siv skein sp800_108 sp800_56c srp6 stateful_rng stream system_rng threefish threefish_avx2 tiger tls tls_cbc tss twofish utils whirlpool x509 x919_mac xmss xtea xts
   INFO: Assuming CPU is little endian
   INFO: Assuming unaligned memory access works
   INFO: Using symlink to link files into build dir (use --link-method to change)
   INFO: Writing amalgamation header to botan_all.h
   INFO: Writing amalgamation source to botan_all.cpp
   INFO: Botan 2.0.1 (VC unknown) (unreleased undated) build setup is complete

I add -march=skylake  to CXXFLAGS in resulting Makefile and use g++ 6.3.0

make
./botan-test
make install
export LD_LIBRARY_PATH=/usr/local/botan/lib:$LD_LIBRARY_PATH
ldconfig -v
cd /usr/local/botan/bin

./botan cpuid
CPUID flags: sse2 ssse3 sse41 sse42 avx2 rdtsc bmi2 adx aes_ni clmul rdrand rdseed

./botan speed --msec=3000 --buf-size=16384 AES-128/GCM
AES-128/GCM(16) [clmul] 353356 key schedule/sec; 0.00 ms/op 11314 cycles/op (2 ops in 0 ms)
AES-128/GCM(16) [clmul] 3638268 iv setup/sec; 0.00 ms/op 1107.25 cycles/op (184408 ops in 50.69 ms)
AES-128/GCM(16) [clmul] encrypt 481.408 MiB/sec 7.94 cycles/byte (1440.69 MiB in 2992.65 ms)
AES-128/GCM(16) [clmul] decrypt 480.228 MiB/sec 7.96 cycles/byte (1440.69 MiB in 3000.01 ms)

./botan speed --msec=3000 --buf-size=16384 AES-192/GCM
AES-192/GCM(16) [clmul] 353544 key schedule/sec; 0.00 ms/op 11277 cycles/op (2 ops in 0 ms)
AES-192/GCM(16) [clmul] 3593550 iv setup/sec; 0.00 ms/op 1107.70 cycles/op (180384 ops in 50.20 ms)
AES-192/GCM(16) [clmul] encrypt 473.789 MiB/sec 8.07 cycles/byte (1409.25 MiB in 2974.43 ms)
AES-192/GCM(16) [clmul] decrypt 469.746 MiB/sec 8.14 cycles/byte (1409.25 MiB in 3000.02 ms)

./botan speed --msec=3000 --buf-size=16384 AES-256/GCM
AES-256/GCM(16) [clmul] 322788 key schedule/sec; 0.00 ms/op 12400 cycles/op (2 ops in 0 ms)
AES-256/GCM(16) [clmul] 3601824 iv setup/sec; 0.00 ms/op 1120.54 cycles/op (177804 ops in 49.36 ms)
AES-256/GCM(16) [clmul] encrypt 464.090 MiB/sec 8.24 cycles/byte (1389.09 MiB in 2993.15 ms)
AES-256/GCM(16) [clmul] decrypt 463.026 MiB/sec 8.26 cycles/byte (1389.09 MiB in 3000.03 ms)

To complete the comparison with my AES-GCM slide here is the performance of ChaCha20Poly1305.

./botan speed --msec=3000 --buf-size=16384 ChaCha20Poly1305
ChaCha20Poly1305 [base] 263365 key schedule/sec; 0.00 ms/op 13292 cycles/op (2 ops in 0 ms)
ChaCha20Poly1305 [base] 2133411 iv setup/sec; 0.00 ms/op 1880.05 cycles/op (200086 ops in 93.79 ms)
ChaCha20Poly1305 [base] encrypt 521.053 MiB/sec 7.34 cycles/byte (1563.17 MiB in 3000.02 ms)
ChaCha20Poly1305 [base] decrypt 521.454 MiB/sec 7.33 cycles/byte (1563.17 MiB in 2997.72 ms)

- So, instead of 0.65 cpb with OpenSSL we only get 7.94 cpb with Botan for AES-128-GCM encryption. More than factor 12! Am I doing something wrong?
- ChaCha20Poly1305 with 7.34 cpb is faster than Botan's AES-GCM, but much slower than OpenSSL's 1.7cpb.
- BTW, I think that AES-128/GCM(16)[clmul] is note quite complete as information, because it uses AES-NI AND PCLMUL, not only PCLMUL as clmul indicates.
- I know that benchmarking is tricky. Am I doing something wrong here? Are we comparing apples and pears? Perhaps my other issue #968 helps comparing OpenSSL and Botan results in the future.

**Comment by SideChannel** [comment]
Just for completeness, here is the output of 

./botan speed --msec=3000 --buf-size=16384 > botan-speed.log.txt 2>&1
and
./openssl speed > openssl-speed.log.txt

Same machine, same compiler.

[botan-speed.log.txt](https://github.com/randombit/botan/files/890498/botan-speed.log.txt)
[openssl-speed.log.txt](https://github.com/randombit/botan/files/890499/openssl-speed.log.txt)

**Comment by randombit** [comment]
The difference is is likely due to the fact that in Botan (currently) AES and GCM are independently optimized for AES-NI/CLMUL instructions, but there is no combined `AES_GCM` `Cipher_Mode` class that specifically accelerates AES+GCM. This means for example each counter block is encrypted with AES-NI in `xmm` registers, then saved to a byte buffer, XORed using the 'standard' byte at a time loop in gcm.cpp, then reloaded into another `xmm` register for the accelerated `clmul` handling. IIRC OpenSSL special cases AES+GCM so that data never leaves `xmm` registers during processing, which is going to be much faster.

Implementing it as OpenSSL does it is the 'stiched' mode referred to in #866

**Comment by matejk** [comment]
Is someone already planning to implement the 'stiched' mode to improve performance?

I am willing to bite in this however I might need a bit of assistance.

**Comment by randombit** [comment]
@matejk Not that I know of and patch for this would be most welcome. I've sent you an invite to the Botan developer Slack as another channel for quick questions.

**Comment by randombit** [comment]
One major reason for the performance difference turned out to be because we were using an inefficient algorithm for clmul reductions. Using a 4 block wide reduction (added in eab327de) substantially improves bulk throughput. On Skylake i7-6700k, OpenSSL 1.1.0g vs Botan master.

1k buffer:

OpenSSL: 4643 MiB/s
Botan: 1230 MiB/s (3.2 cycles per byte)
OpenSSL advantage: 3.7x

16k buffers

OpenSSL: 5953 MiB/s
Botan: 2342 MiB/s (1.6 cycles per byte)
OpenSSL advantage: 2.5x

Likely much of the remaining advantage is due to botan not having a stitched implementation.

**Comment by randombit** [comment]
I was looking at the ChaCha20Poly1305 results. OpenSSL has AVX2 codepaths for both ChaCha20 and Poly1305, whereas Botan has only SSE2 for ChaCha20 and ALU for Poly1305. An 8x AVX2 path for ChaCha20 and a SSE2/AVX2 codepath for Poly1305 should get us within striking range of OpenSSL perf.

**Comment by randombit** [comment]
With #1662 adding AVX2 support to ChaCha, I'm now seeing 4.1 cycles per byte for 16K ChaCha20Poly1305. Which is much worse than OpenSSL's 1.7 cpb, but much better than previous result of 7.3.

**Comment by randombit** [comment]
My measurements are also on 4 GHz i7-6700k, HT and turboboost disabled, 16K buffers.

All numbers are cycles per byte. In reported/current (N.Nx) gives the ratio
between that result and OpenSSL.


Algo             | OpenSSL | Reported 2.0.1     | Current 2.8.0     | Improvement
-----------------|---------|---------------|--------------|-------------
AES-128/GCM      | .65     | 7.94 (12.2x)  | 1.44 (2.2x)  | 5.5x
AES-192/GCM      | .77     | 8.07 (10.5x)  | 1.56 (2x)    | 5.2x
AES-256/GCM      | .90     | 8.24 (9.1x)   | 1.7 (1.8x)   | 4.8x
ChaCha20Poly1305 | 1.7     | 7.34 (4.3x)   | 4.07 (2.4x)  | 1.8x

**Comment by randombit** [comment]
Another followup using Botan 3.11.1, same machine as above comments, 16K buffers

AES-128/GCM: 1.16 cycles per byte
AES-192/GCM: 1.23 cycles per byte
AES-256/GCM: 1.38 cycles per byte
ChaCha20Poly1305: 2.66 cycles per byte

Checking around, the only implementations I can find in common libraries that are faster than these numbers are OpenSSL and libgcrypt, both of which used combined AES+GCM codepaths.

Potentially of note, Botan's AES/GCM is these days faster OpenSSL 3.5 for small messages, which causes me to be skeptical of the numbers quoted in the slide in OP.

OpenSSL 3.6.0 (`openssl speed -seconds 1 -evp aes-128-gcm`)

> type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
> AES-128-GCM      73540.59k   283567.77k   942306.56k  2543456.26k  5307768.83k  5724897.28k

[Note also how these numbers are *much* worse also than OP quotes as directly the output of the same command using OpenSSL 1.1.0!]

Botan 3.11.1 with better performance in bytes/sec, but reporting cpb much worse than the slide:

```
AES-128/GCM(16) [clmul] encrypt buffer size 16 bytes: 107.584 MiB/sec 35.53 cycles/byte (53.81 MiB in 500.19 ms)
AES-128/GCM(16) [clmul] encrypt buffer size 64 bytes: 413.402 MiB/sec 9.25 cycles/byte (206.75 MiB in 500.12 ms)
AES-128/GCM(16) [clmul] encrypt buffer size 256 bytes: 1063.421 MiB/sec 3.59 cycles/byte (531.75 MiB in 500.04 ms)
```

So in summary, these days Botan's AEAD is much faster, OpenSSL's is much slower, and now their relative performance is similar. :rocket: Closing.

**Comment by port19x** [comment]
@randombit I can't follow your latest update.
If you have the time, could you make it a table like your last comment in 2018?

If Botan is faster and Openssl is slower, how are their relative performance similar?
What are the cycles per byte for openssl 3.6.0 (or newer)?

If it matters, I care about small messages with AES-256-GCM AEAD.
pystardust/ani-cli#1779

**Comment by randombit** [comment]
@port19x I meant by that comment that Botan has improved relative to the performance of Botan 2.0, and similarly OpenSSL has regressed in performance since OpenSSL 1.1.0, to the point that the performance ranges now overlap. For small messages *on my machine* most recent Botan is faster than OpenSSL. For large messages (> 4K) OpenSSL is still faster, at least on a Skylake processor. It would be worthwhile checking on your particular target hardware though, since for both the main question is how effectively the implementation uses available SIMD and AES support instructions. (For example Aarch64 likely has very different results for both and the numbers quoted here may not be representative of either implementation)

**Comment by port19x** [comment]
Thank you, that helps

**Comment by SideChannel** [comment]
@port19x @randombit

Hi, as I am the original poster of this issue, I would like to add my 2 cent.
First, congratulations to Jack, Botan improved a lot since my posting in 2017. Not only in performance of AEAD.

Let's now come to the performance question. We all know, that OpenSSL decreased in performance with the step to 3.0. This improved somewhat with newer versions, but it still hasn't reached the old performance of 1.1.x. 

`Potentially of note, Botan's AES/GCM is these days faster OpenSSL 3.5 for small messages, which causes me to be skeptical of the numbers quoted in the slide in OP.` randombit's note

I do not have access to my old Skylake machine anymore. Instead we have

- AMD Ryzen 9 7950X3D@4.2GHz processor with 128 GB RAM
- OpenSuSE Tumbleweed Linux, gcc 16.1
- compilation of OpenSSL 4.0.1
- tar xvf openssl-4.0.1.tar.gz
  cd openssl-4.0.1
  CC=gcc-16 ./Configure -march=znver4 -mtune=znver4 --prefix=/usr/local/ssl enable-ec_nistp_64_gcc_128
  make; make test; make install
  cd /usr/local/ssl/bin
  export LD_LIBRARY_PATH=/usr/local/ssl/lib:$LD_LIBRARY_PATH
 - ./openssl speed -evp aes-128-gcm 
  gives
Doing AES-128-GCM ops for 3s on 16 size blocks: 38783317 AES-128-GCM ops in 3.00s
Doing AES-128-GCM ops for 3s on 64 size blocks: 37100416 AES-128-GCM ops in 3.00s
Doing AES-128-GCM ops for 3s on 256 size blocks: 32501067 AES-128-GCM ops in 3.00s
Doing AES-128-GCM ops for 3s on 1024 size blocks: 19240492 AES-128-GCM ops in 3.00s
Doing AES-128-GCM ops for 3s on 8192 size blocks: 5630938 AES-128-GCM ops in 3.00s
Doing AES-128-GCM ops for 3s on 16384 size blocks: 3117573 AES-128-GCM ops in 3.00s
version: 4.0.1
built on: Wed Jun 24 09:23:46 2026 UTC
options: bn(64,64)
compiler: gcc-16 -fPIC -pthread -m64 -Wa,--noexecstack -Wall -O3 -march=znver4 -mtune=znver4 -DL_ENDIAN -DOPENSSL_PIC -DNDEBUG
CPUINFO: OPENSSL_ia32cap=0x7ed8320b078bffff:0x00405fdef1bf97a9:0x0000002010000010:0x0000000000000000:0x0000000000000000
The 'numbers' are in 1000s of bytes per second processed.
type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
AES-128-GCM     206844.36k   791475.54k  2773424.38k  6567421.27k 15376214.70k 17026105.34k
- Likewise we have
AES-192-GCM     193479.80k   746371.24k  2623779.67k  6231268.35k 14426808.32k 15834606.25k
AES-256-GCM     188385.31k   731355.46k  2560223.57k  6042543.79k 13450007.89k 14752568.66k
- How do I obtain the cycles per byte in my slide? By dividing the Frequency through the output of openssl speed, i.e., 4.2e9/17026105.34e3.
- This gives the following slide. Note that the performance of the internal AES-NI commands improved a lot since Skylake. Therefore, the theoretical maximum with AES-128 is now 0.25 instead of 0.66 with Skylake. Note that I do not know of any current crypto lib which uses the newer VAES* commands.

<img width="1357" height="705" alt="Image" src="https://github.com/user-attachments/assets/43bf491c-9121-42f9-8fcc-7a7504902b18" />

Let's now come to Botan 3.12.0.
 tar xvf botan-3.12.0.tar.xz
 ./configure.py --amalgamation  --prefix=/usr/local/botan
 add -march=znver4 to CXXFLAGS in Makefile

 make
 make check
 make install
 export LD_LIBRARY_PATH=/usr/local/botan/lib:$LD_LIBRARY_PATH
 cd /usr/local/botan/bin
 ./botan cpuid
CPUID flags: sse2 ssse3 avx2 avx512 rdtsc adx bmi2 gfni rdrand rdseed aesni clmul intel_sha avx2_aes avx512_aes avx2_clmul avx512_clmul
 for i in 16 64 256 1024 8192 16384
 do
 ./botan speed --msec=3000 --buf-size=$i AES-128/GCM
 done
AES-128/GCM(16) [avx512_clmul] 120758 key schedule/sec; 0.01 ms/op 33810 cycles/op (2 ops in 0 ms)
AES-128/GCM(16) [avx512_clmul] encrypt buffer size 16 bytes: 72.897 MiB/sec 54.84 cycles/byte (218.75 MiB in 3000.81 ms)
AES-128/GCM(16) [avx512_clmul] decrypt buffer size 16 bytes: 68.365 MiB/sec 58.48 cycles/byte (205.12 MiB in 3000.44 ms)
AES-128/GCM(16) [avx512_clmul] 166917 key schedule/sec; 0.01 ms/op 24423 cycles/op (2 ops in 0 ms)
AES-128/GCM(16) [avx512_clmul] encrypt buffer size 64 bytes: 290.783 MiB/sec 13.75 cycles/byte (872.38 MiB in 3000.09 ms)
AES-128/GCM(16) [avx512_clmul] decrypt buffer size 64 bytes: 268.794 MiB/sec 14.87 cycles/byte (806.44 MiB in 3000.21 ms)
AES-128/GCM(16) [avx512_clmul] 170619 key schedule/sec; 0.01 ms/op 23919 cycles/op (2 ops in 0 ms)
AES-128/GCM(16) [avx512_clmul] encrypt buffer size 256 bytes: 734.614 MiB/sec 5.44 cycles/byte (2203.88 MiB in 3000.04 ms)
AES-128/GCM(16) [avx512_clmul] decrypt buffer size 256 bytes: 1017.481 MiB/sec 3.93 cycles/byte (3052.50 MiB in 3000.06 ms)
AES-128/GCM(16) [avx512_clmul] 174034 key schedule/sec; 0.01 ms/op 23478 cycles/op (2 ops in 0 ms)
AES-128/GCM(16) [avx512_clmul] encrypt buffer size 1024 bytes: 3272.751 MiB/sec 1.22 cycles/byte (9818.31 MiB in 3000.02 ms)
AES-128/GCM(16) [avx512_clmul] decrypt buffer size 1024 bytes: 2948.620 MiB/sec 1.36 cycles/byte (8845.88 MiB in 3000.01 ms)
AES-128/GCM(16) [avx512_clmul] 160346 key schedule/sec; 0.01 ms/op 25116 cycles/op (2 ops in 0 ms)
AES-128/GCM(16) [avx512_clmul] encrypt buffer size 8192 bytes: 9231.426 MiB/sec 0.43 cycles/byte (27694.31 MiB in 3000.00 ms)
AES-128/GCM(16) [avx512_clmul] decrypt buffer size 8192 bytes: 8401.987 MiB/sec 0.48 cycles/byte (25206.00 MiB in 3000.00 ms)
AES-128/GCM(16) [avx512_clmul] 163492 key schedule/sec; 0.01 ms/op 25053 cycles/op (2 ops in 0 ms)
AES-128/GCM(16) [avx512_clmul] encrypt buffer size 16384 bytes: 10648.435 MiB/sec 0.38 cycles/byte (31945.31 MiB in 3000.00 ms)
AES-128/GCM(16) [avx512_clmul] decrypt buffer size 16384 bytes: 9522.601 MiB/sec 0.42 cycles/byte (28567.81 MiB in 3000.00 ms)

Similar for AES-192/GCM and AES-256/GCM, as well as ChaCHa20Poly1305. This give the following slide

<img width="1333" height="614" alt="Image" src="https://github.com/user-attachments/assets/636e7654-446c-46ee-bfc7-b3fc7064d9d6" />

Same machine, the only difference is the usage of gcc 15.3 (I forgot to change this). 
- So. What do we learn? 
- OpenSSL got slower. A lot slower, especially for small files, up to, say, 256 byte.
- OpenSSL reaches the asymptotic limit for large files, e.g., 0.25 cpb for 16384 byte messages and AES-128.
- Botan's AEAD performance much improved, e.g. 0.38 cpb (versus 0.25 of OpenSSL). So the factor is now only 1.5. This is absolutely okay for me and most applications. Note we are coming from a factor of above 12! Great news!
- I cannot confirm that Botan is faster than OpenSSL, even for small messages. There we have 54.84 cpb Botan versus 20.31 cpb OpenSSL. Factor 20.7.
- Note that benchmarking for small messages is an art. There is all this small overhead here and there. Further note that I did not change the BIOS for benchmarking. So, no overclocking, only EXPO memory setting. No changes to HT and Turboboost.
- Differences to your benchmarks: I use speed for 3 seconds as is custom in OpenSSL, I do not put of HT and Turboboost!
- The benchmarking conditions for OpenSSL and Botan were the same. I hope this convinces you @randombit that my numbers are at least consistent.
- Here are the full results

[botan-benchmark.txt](https://github.com/user-attachments/files/30259325/botan-benchmark.txt)

[linux_benchmark.Ryzen7950X3D.4.0.1.txt](https://github.com/user-attachments/files/30259429/linux_benchmark.Ryzen7950X3D.4.0.1.txt)
- If Botans AES-128-GCM-SIV becomes faster than OpenSSL's (currently 4.99 cpb for 16384 byte messages instead of the 0.98 cpb announced by the authors of AES-GCM-SIV) and HPKE runs with PQC, than I promise to change to Botan asap. :-) Just a joke, I use different crypto libs depending on the (embedded) use case. The documentation of Botan is far better than OpenSSL's (but this is fairly easy).
